metaflow/cmd/diff/__init__.py

```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+import os
+import shutil
+import sys
+from subprocess import PIPE, run
+from tempfile import TemporaryDirectory
+
+from metaflow._vendor import click
+from metaflow import Run, namespace
+from metaflow.cli import echo_always
+
+EXCLUSIONS = [
+    "metaflow/",
+    "metaflow_extensions/",
+    "INFO",
+    "CONFIG_PARAMETERS",
+    "conda.manifest",
+]
+
+
+def echo(line):
+    echo_always(line, err=True, fg="magenta")
+
+
+def extract_code_package(runspec, exclusions):
+    try:
+        namespace(None)
+        run = Run(runspec)
+        echo(f"✅  Run *{runspec}* found, downloading code..")
+    except:
+        echo(f"❌  Run **{runspec}** not found")
+        sys.exit(1)
+
+    if run.code is None:
+        echo(
+            f"❌  Run **{runspec}** doesn't have a code package. Maybe it's a local run?"
+        )
+        sys.exit(1)
+
+    tar = run.code.tarball
+    members = [
+        m for m in tar.getmembers() if not any(m.name.startswith(x) for x in exclusions)
+    ]
+
+    tmp = TemporaryDirectory()
+    tar.extractall(tmp.name, members)
+    return tmp
+
+
+def perform_diff(source_dir, target_dir=None, output=False):
+    if target_dir is None:
+        target_dir = os.getcwd()
+
+    diffs = []
+    for dirpath, dirnames, filenames in os.walk(source_dir):
+        for fname in filenames:
+            # NOTE: the paths below need to be set up carefully
+            # for the `patch` command to work. Better not to touch
+            # the directories below. If you must, test that patches
+            # work after you changes.
+            #
+            # target_file is the git repo in the current working directory
+            rel = os.path.relpath(dirpath, source_dir)
+            target_file = os.path.join(rel, fname)
+            # source_file is the run file loaded in a tmp directory
+            source_file = os.path.join(dirpath, fname)
+
+            if sys.stdout.isatty() and not output:
+                color = ["--color"]
+            else:
+                color = ["--no-color"]
+
+            if os.path.exists(os.path.join(target_dir, target_file)):
+                cmd = (
+                    ["git", "diff", "--no-index", "--exit-code"]
+                    + color
+                    + [
+                        target_file,
+                        source_file,
+                    ]
+                )
+                result = run(cmd, text=True, stdout=PIPE, cwd=target_dir)
+                if result.returncode == 0:
+                    echo(f"✅ {target_file} is identical, skipping")
```

Comment:
Minor nit: you have 2 spaces after the emojis elsewhere but 1 here and on line 92.